app/javascript/components/Collaborators/Form.tsx

```diff
@@ -79,138 +58,169 @@ const CollaboratorForm = () => {
     return !product.has_another_collaborator && product.published;
   };
 
-  const [products, setProducts] = React.useState<CollaboratorProduct[]>(() =>
-    formData.products.map((product) =>
-      isEditing
-        ? {
-            ...product,
-            percent_commission: product.percent_commission || defaultPercentCommission.value,
-            dont_show_as_co_creator: applyToAllProducts ? dontShowAsCoCreator : product.dont_show_as_co_creator,
-            has_error: false,
-          }
-        : {
-            ...product,
-            enabled: shouldEnableProduct(product),
-            percent_commission: defaultPercentCommission.value,
-            has_error: false,
-          },
-    ),
-  );
+  const form = useForm<{
+    email: string;
+    apply_to_all_products: boolean;
+    percent_commission: number | null;
+    dont_show_as_co_creator: boolean;
+    products: CollaboratorProduct[];
+    default_commission_has_error: boolean;
+  }>({
+    email: isEditing ? formData.email : "",
+    apply_to_all_products: initialApplyToAllProducts,
+    percent_commission: initialDefaultPercentCommission,
+    dont_show_as_co_creator: initialDontShowAsCoCreator,
+    products: formData.products.map((product): CollaboratorProduct => {
+      if (isEditing) {
+        return {
+          ...product,
+          percent_commission: product.percent_commission || initialDefaultPercentCommission,
+          dont_show_as_co_creator: initialApplyToAllProducts
+            ? initialDontShowAsCoCreator
+            : product.dont_show_as_co_creator,
+          has_error: false,
+        };
+      }
+      return {
+        ...product,
+        enabled: shouldEnableProduct(product),
+        percent_commission: initialDefaultPercentCommission,
+        has_error: false,
+        dont_show_as_co_creator: false,
+      };
+    }),
+    default_commission_has_error: false,
+  });
 
-  const productsWithAffiliates = products.filter((product) => product.enabled && product.has_affiliates);
+  const { data, setData, processing, errors } = form;
+
+  const productsWithAffiliates = data.products.filter((product) => product.enabled && product.has_affiliates);
   const listedProductsWithAffiliatesCount =
     productsWithAffiliates.length <= MAX_PRODUCTS_WITH_AFFILIATES_TO_SHOW + 1
       ? productsWithAffiliates.length
       : MAX_PRODUCTS_WITH_AFFILIATES_TO_SHOW;
 
   const handleProductChange = (id: string, attrs: Partial<CollaboratorProduct>) => {
-    setProducts((prevProducts) =>
-      prevProducts.map((item) => (item.id === id ? { ...item, ...attrs, has_error: false } : item)),
+    setData(
+      "products",
+      data.products.map((item) => (item.id === id ? { ...item, ...attrs, has_error: false } : item)),
```

Comment:
Can you use Inertia's form errors to handle errors? https://inertiajs.com/docs/v2/the-basics/forms#form-errors

Same comments applies to a lot of parts of this code, it would make it way more simple and easier to reason about.